src/primitives/iter.rs

```diff
@@ -0,0 +1,441 @@
+// SPDX-License-Identifier: MIT
+
+//! Iterator Adaptors
+//!
+//! This module provides iterator adaptors that can be used to verify and generate checksums, HRP
+//! strings, etc., in a variety of ways, without any allocations.
+//!
+//! In general, directly using these adaptors is not very ergonomic, and users are recommended to
+//! instead use the higher-level functions at the root of this crate.
+//!
+//! # Examples
+//!
+//! ```rust
+//! use bech32::{Bech32, ByteIterExt, Fe32IterExt, Fe32, Hrp};
+//!
+//! let witness_prog = [
+//!     0x75, 0x1e, 0x76, 0xe8, 0x19, 0x91, 0x96, 0xd4,
+//!     0x54, 0x94, 0x1c, 0x45, 0xd1, 0xb3, 0xa3, 0x23,
+//!     0xf1, 0x43, 0x3b, 0xd6,
+//! ];
+//! let hrp = Hrp::parse_unchecked("bc");
+//! let witness_version = Fe32::Q; // Witness version 0.
+//! let iterator = witness_prog
+//!     .iter()
+//!     .copied()
+//!     .byte_to_fe32_iter()
+//!     .checksummed_iter::<Bech32>()
+//!     .hrpstring_iter(&hrp, Some(witness_version));
+//! let hrpstring: String = iterator.collect();
+//! assert_eq!(hrpstring.to_uppercase(), "BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4");
+//! ```
+
+use crate::primitives::checksum::{self, Checksum, PackedFe32};
+use crate::primitives::gf32::Fe32;
+use crate::primitives::hrp::{self, Hrp};
+
+/// Extension trait for byte iterators which provides an adaptor to GF32 elements.
+pub trait ByteIterExt: Sized + Iterator<Item = u8> {
+    /// Adapts the byte iterator to output GF32 field elements instead.
+    fn byte_to_fe32_iter(mut self) -> ByteToFe32Iter<Self> {
+        ByteToFe32Iter { last_byte: self.next(), bit_offset: 0, iter: self }
+    }
+}
+
+impl<I> ByteIterExt for I where I: Iterator<Item = u8> {}
+
+/// Extension trait for field element iterators.
+pub trait Fe32IterExt: Sized + Iterator<Item = Fe32> {
+    /// Adapts the `Fe32` iterator to output bytes instead.
+    ///
+    /// If the total number of bits is not a multiple of 8, any trailing bits
+    /// are simply dropped.
+    fn fe32_to_byte_iter(mut self) -> Fe32ToByteIter<Self> {
+        Fe32ToByteIter { last_fe: self.next(), bit_offset: 0, iter: self }
+    }
+
+    /// Adapts the Fe32 iterator to append a checksum to the end of the data.
+    fn checksummed_iter<Ck: Checksum>(self) -> ChecksummedIter<Self, Ck> {
```

Comment:
In e2527dc19c3e8e9a4cd678a955fc5b161e175c87:

These shouldn't have `iter` suffices if they're iterator adaptors.

Reply:
I'd also make `fe32` and `byte` plural in both conversion adaptor names.

Reply:
Or even do `fe32string_to_bytestring` and `bytestring_to_fe32string` to emphasize that it's not individual characters being converted.

Reply:
Cool, I elected for the following iterator types and methods that construct them
- `BytesToFes` and `bytes_to_fes`
- `FesToBytes` and `fes_to_bytes`
- `Checksummed` and `checksummed`
- `HrpstringChars` and `hrpstring_chars`

I feel like the last one could be better named, example usage:
```rust
    let iter = data.as_ref().iter().copied();
    match variant {
        Variant::Bech32 =>
            for c in iter.checksummed::<Bech32>().hrpstring_chars(&hrp, None) {
                fmt.write_char(c)?;
            },
        Variant::Bech32m =>
            for c in iter.checksummed::<Bech32m>().hrpstring_chars(&hrp, None) {
                fmt.write_char(c)?;
            },
    }
```

Reply:
And for more stimulation
```rust
let addr = witness_prog
    .iter()
    .copied()
    .bytes_to_fes()
    .checksummed::<Bech32>()
    .hrpstring_chars(&hrp, Some(witness_version))
    .collect::<String>()
```

Reply:
All sounds good! I think `hrpstring_chars` is fine. `encoded` doesn't give any clue that the iterator adaptor is adding an HRP and a checksum.

Reply:
Ah, I see, we've already called `hrp_checksummed` at this point. I think just `.chars()` is fine for this.

Reply:
Actually I think we should invert this, and iterate over `char` by default, and require an adaptor to get the `fe`s out. We can call the latter `fes_for_checksum_engine`. Fine to have a mouthful of a name because it's not likely to be a commonly-used adaptor.

Reply:
In 51781bc I'd still like to change this.

Reply:
In d69ea01 I'd still like to change this.

Reply:
hmm, interesting, just saw this after - will play with it now.